src/main/java/org/folio/auth/authtokenmodule/tokens/Token.java

```diff
@@ -251,30 +257,56 @@ protected void validateCommon(HttpServerRequest request) throws TokenValidationE
       }
 
       if (request == null) {
-        return;
+        promise.complete(this);
+        return promise.future();
       }
 
       // Check that some items in the headers match what are in the token.
-      String headerTenant = request.headers().get(XOkapiHeaders.TENANT);
-      if (!claims.getString("tenant").equals(headerTenant)) {
-        throw new TokenValidationException("Tenant mismatch: tenant in header does not equal tenant in token", 403);
-      }
       String headerUserId = request.headers().get(XOkapiHeaders.USER_ID);
       String claimsUserId = claims.getString("user_id");
       if (headerUserId != null && claimsUserId != null && !claimsUserId.equals(headerUserId)) {
         throw new TokenValidationException("User id in header does not equal user id in token", 403);
       }
+
+      String headerTenant = request.headers().get(XOkapiHeaders.TENANT);
+      if (!claims.getString("tenant").equals(headerTenant)) {
+        isCrossTenantRequest(context)
+          .compose(aResult -> Boolean.TRUE.equals(aResult) ? Future.succeededFuture(this)
+            : Future.failedFuture(new TokenValidationException(TENANT_MISMATCH_EXCEPTION_MESSAGE, 403)))
+          .onComplete(promise);
+      } else {
+        promise.complete(this);
```

Comment:
I think we still need the check of `claims.getString("tenant").equals(headerTenant)` don't we when it is not a cross tenant request? Otherwise a token could be used to request another tenant.

Reply:
I believe this code branch is for when there is a tenant id match. check if `claims.getString("tenant").equals(headerTenant)` is already done at the beginning of the if for every request. If there is a tenant mismatch and it is not a cross tenant request, this failed future is returned `Future.failedFuture(new TokenValidationException(TENANT_MISMATCH_EXCEPTION_MESSAGE, 403))` which would yield the same exception as the OLD code that was replaced.